cEP-0026.md

```diff
@@ -0,0 +1,189 @@
+providing the user various options to perform caching according to a project's
+needs.
+
+## Implementation
+
+The implementation will be divided into four parts
+
+- Prototype for `FileFactory` class implementation.
+- Prototype for `Directory` class implementation.
+- Ignore directories functionality.
+- Cache control flags.
+
+### Prototype of FileFactory class implementation
+
+This class will be implemented according to the Factory design pattern and will
+act as a proxy to dealing with files. This class will have various methods to
+provide the file contents in different forms for e.g. string, binary
+format, etc. The biggest benefit that we will get out of this class would be
+that we can replace the actual file contents in our file-dict with these
+objects which will itself be a performance improvement.
+
+The following is the prototype for `FileFactory` class.
+
+```python
+import linecache
+import os
+import pickle
+
+from coalib.core.PersistentHash import persistent_hash
+
+
+class FileFactory:
+    def __init__(self, filename, parent):
+        self.filename = filename # absolute file path
+        self.parent = os.path.abspath(os.path.join(p, os.pardir)) # parent dir
+        self._content = open(self.filename)
+
+    def open(filename, newline=None):
+        return FileFactory(filename, newline=newline)
+
+    def timestamp(self):
+        # get the last modified timestamp of the file
+        return os.path.gettime(self.path)
+
+    def file_as_string(self, encoding=None):
+        return "".join(self._content)
+
+    def file_as_bytes(self):
+        # return the pickled version of the self._content
+        pickle.dumps(self.filename)
+
+    def hash(self):
+        # hash the absolute file path to get a unique hash
+        return persistent_hash(self.filename)
+```
+
+### Prototype of Directory class implementation
+
+The objects of this class will also reside in the file-dict (need to think of
+a better name for the file-dict since it will no longer contain any files).
+The objects of this class can also be used by `DirectoryBears` as discussed
+on this thread [#4999](https://github.com/coala/coala/issues/4999) by performing
+a simple type check on the entries in the dict, whether the objects belong to
+`FileFactory` or `Directory`.
+
+Following is the prototype for the `Directory` class.
+
+```python
+import os
+
+from coalib.parsing.Globbing import glob
+
+
+class DirectoryFactory:
+    def __init__(self, path):
+        self.path = path
+        self.parent = os.path.abspath(os.path.join(p, os.pardir)) # parent dir
+        self.files = 0 # number of files contained in a directory
+        self.subdirs = 0 # number of sub-directories contained in a directory
+
+    def count_children(self):
+        self.children = os.listdir(self.path)
+
+        for child in self.children:
+            if os.path.isfile(child):
+                self.files += 1
+
+        self.subdirs = len(self.children) - self.files
```

Comment:
Do you have a usecase for `count_children`?

Reply:
Not really its only used to calculate the subdirs. I guess you are right this shouldn't be an attribute of the class but just a temporary variable so I will keep `children` but remove `self` from it.